src/v/redpanda/admin_server.cc

```diff
@@ -2898,6 +2899,79 @@ void admin_server::register_transaction_routes() {
       });
 }
 
+void admin_server::register_self_test_routes() {
+    register_route<user>(
+      ss::httpd::debug_json::self_test_start,
+      [this](std::unique_ptr<ss::httpd::request>)
+        -> ss::future<ss::json::json_return_type> {
+          auto r = co_await _self_test_frontend.invoke_on(
+            cluster::self_test_frontend::shard, [](auto& self_test_frontend) {
+                return self_test_frontend.start_test();
+            });
+          if (std::holds_alternative<cluster::self_test_status_report>(r)) {
+              const auto& st = std::get<cluster::self_test_status_report>(r);
+              throw ss::httpd::base_exception(
+                fmt::format("Failed to start self test, reason: {}", st),
+                ss::httpd::reply::status_type::internal_server_error);
+          }
+          const auto& test_id = std::get<uuid_t>(r);
+          vlog(
+            logger.info, "Request to start self test succeeded: {}", test_id);
+          co_return ss::json::json_return_type(test_id);
+      });
+
+    register_route<user>(
+      ss::httpd::debug_json::self_test_stop,
+      [this](std::unique_ptr<ss::httpd::request>)
+        -> ss::future<ss::json::json_return_type> {
+          auto r = co_await _self_test_frontend.invoke_on(
+            cluster::self_test_frontend::shard, [](auto& self_test_frontend) {
+                return self_test_frontend.stop_test();
+            });
+          vlog(logger.info, "Request to stop self test, status: {}", r);
+          co_return ss::json::json_void();
+      });
+
+    register_route<user>(
+      ss::httpd::debug_json::self_test_status,
+      [this](std::unique_ptr<ss::httpd::request>)
+        -> ss::future<ss::json::json_return_type> {
+          auto status = co_await _self_test_frontend.invoke_on(
+            cluster::self_test_frontend::shard, [](auto& self_test_frontend) {
+                return self_test_frontend.status().then(
+                  [&self_test_frontend](auto cur) {
+                      /// If the current test has completed, return last cached
+                      /// result set to display to user
+                      return cur.is_test_active()
+                               ? cur
+                               : self_test_frontend.previous_run();
+                  });
+            });
+
+          using result_t = ss::httpd::debug_json::self_test_node_status;
+          ss::httpd::debug_json::self_test_response r;
```

Comment:
nit: could reserve the r.reports vector size ahead of loop

Reply:
Hmm seems like the json::json_list object doesn't have a reserve method